ksql-rest-app/src/main/java/io/confluent/ksql/rest/util/ClusterTerminator.java

```diff
@@ -0,0 +1,161 @@
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.metastore.StructuredDataSource;
+import io.confluent.ksql.parser.tree.AbstractStreamDropStatement;
+import io.confluent.ksql.parser.tree.DropStream;
+import io.confluent.ksql.parser.tree.DropTable;
+import io.confluent.ksql.parser.tree.QualifiedName;
+import io.confluent.ksql.rest.server.KsqlRestConfig;
+import io.confluent.ksql.rest.server.computation.Command;
+import io.confluent.ksql.rest.server.computation.CommandId;
+import io.confluent.ksql.serde.DataSource.DataSourceType;
+import io.confluent.ksql.util.ExecutorUtil;
+import io.confluent.ksql.util.KafkaTopicClient;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import io.confluent.ksql.util.QueryMetadata;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.producer.Producer;
+
+public class ClusterTerminator {
+
+  @SuppressWarnings("unchecked")
+  public void terminateCluster(
+      final KsqlConfig ksqlConfig,
+      final KsqlEngine ksqlEngine,
+      final List<String> sourcesList,
+      final boolean isKeep,
+      final Consumer<CommandId, Command> commandConsumer,
+      final Producer<CommandId, Command> commandProducer
+  ) {
+    ksqlEngine.stopAcceptingStatemens();
+    terminateCluster(ksqlEngine, sourcesList, isKeep);
+    // Delete the command topic
+    deleteCommandTopic(ksqlConfig, ksqlEngine, commandConsumer, commandProducer);
+  }
+
+  private void terminateCluster(
+      final KsqlEngine ksqlEngine,
+      final List<String> sourcesList,
+      final boolean isKeep
+  ) {
+    // Terminate all queries
+    getListForSet(ksqlEngine).forEach(
+        queryMetadata -> {
+          if (queryMetadata instanceof PersistentQueryMetadata) {
+            final PersistentQueryMetadata persistentQueryMetadata
+                = (PersistentQueryMetadata) queryMetadata;
+            ksqlEngine.terminateQuery(persistentQueryMetadata.getQueryId(), true);
+          }  else {
+            queryMetadata.close();
+          }
+        }
+    );
+
+    // if we have the explicit list of stream/table to delete.
+    final MetaStore metaStore = ksqlEngine.getMetaStore();
+    if (sourcesList.isEmpty()) {
+      metaStore.getAllStructuredDataSources().forEach((s, structuredDataSource) ->
+          deleteSource(s, structuredDataSource, ksqlEngine));
+    } else {
+      if (isKeep) {
+        metaStore.getAllStructuredDataSources().forEach(
+            (sourceName, structuredDataSource) -> {
+              if (!sourcesList.contains(sourceName)) {
+                deleteSource(sourceName, structuredDataSource, ksqlEngine);
+              }
+            }
+        );
+      } else {
+        sourcesList
+            .forEach(sourceName -> deleteSource(
```

Comment:
Does this throw if one of the sources is unknown? Probably should ignore it? Or validate all exist before starting to process the command?

Reply:
Good point. Added check in the REST end point to validate there.